### ISSUE TITLE
Cannot upgrade to new nixos channel

### ISSUE THREAD

**Issue by rowanG077** [tpwrules/nixos-apple-silicon#156]
I upgraded my nix channel and switch to my new configuration. Now everything that requires graphics gives an `Illegal instruction core dumped`. The source is in the asahi mesa driver:

```
> gdb glxgears
...
Reading symbols from glxgears...
(No debugging symbols found in glxgears)
(gdb) run
Starting program: /nix/store/aw4nbyd3z2k3rqygrpn4lxxqf6qy2jma-mesa-demos-9.0.0/bin/glxgears 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/nix/store/jz866c0q32vx9pag7br7dxrqqxkd0m1x-glibc-2.38-27/lib/libthread_db.so.1".

Program received signal SIGILL, Illegal instruction.
0x0000fffff66b08e0 in driCreateNewScreen2 () from /run/opengl-driver/lib/dri/asahi_dri.so
(gdb) bt
#0  0x0000fffff66b08e0 in driCreateNewScreen2 () from /run/opengl-driver/lib/dri/asahi_dri.so
#1  0x0000fffff7766a3c in dri3_create_screen () from /run/opengl-driver/lib/libGLX_mesa.so.0
#2  0x0000fffff7757060 in AllocAndFetchScreenConfigs () from /run/opengl-driver/lib/libGLX_mesa.so.0
#3  0x0000fffff7758118 in __glXInitialize () from /run/opengl-driver/lib/libGLX_mesa.so.0
#4  0x0000fffff7753760 in glXChooseVisual () from /run/opengl-driver/lib/libGLX_mesa.so.0
#5  0x0000000000402cf0 in make_window.constprop ()
#6  0x0000000000401df4 in main ()
```

**Comment by tpwrules** [comment]
The encoding of `udf #0` is `0x00000000`. Did something go wrong with the file system? Can you verify the relevant store paths have the right hash? Try `nix-store --verify --check-contents`.

**Comment by rowanG077** [comment]
Doesn't seem to be an issue. After some thinking it came back with:

```
> nix-store --verify --check-contents
reading the Nix store...
checking path existence...
checking link hashes...
checking store hashes.
> 
```

Which I expect means everything looks good.

**Comment by tpwrules** [comment]
Yeah, that means the store is okay. Doesn’t exclude the possibility something went wrong building. What Nixpkgs hash are you on?

**Comment by rowanG077** [comment]
nixos-unstable channel at commit `97b17f32362e`

**Comment by zzywysm** [comment]
Perhaps we need to explicitly set `withLibunwind` to false?

https://github.com/NixOS/nixpkgs/commit/73f6621a3713fc01c09d2e237bf1cc877be45cfe

**Comment by rowanG077** [comment]
It's already set to disabled in the override.

**Comment by zzywysm** [comment]
I'm also having mesa issues, specifically when in overlay mode:

https://github.com/tpwrules/nixos-apple-silicon/issues/152

Given the bizarre disassembly above, I wonder if this is more evidence for my llvmPackages theory?

**Comment by rowanG077** [comment]
I'm using `replace` mode.

**Comment by zzywysm** [comment]
According to Janne Grunau in Matrix chat, Fedora Asahi now builds mesa+asahi _with_ libunwind, whereas ALARM built mesa+asahi _without_ it.

This is the build specification file for mesa in Fedora Asahi, which I admit I find confusing:

https://copr-dist-git.fedorainfracloud.org/cgit/@asahi/mesa/mesa.git/tree/mesa.spec?h=f39&id=197e90d54335ef8fe2c47aa9ead5e5d2a8a3b6d4

**Comment by rowanG077** [comment]
The disassembly looks fine using that...

**Comment by tpwrules** [comment]
I still think something machine-specific went wrong when building and the file ended up corrupt. I set my system to use the latest `main` commit of this repo (6e324ab) and the nixpkgs you mention (97b17f32362e). I cannot replicate this issue, nor the faulty disassembly, on my system.

I examined store path `/nix/store/yj2pxkiwyf64xi50y6zy7zpmfa705wd6-mesa-24.0.0-drivers/` which came from derivation `/nix/store/6qamlsrgfqzgl5k134bvvr6xic4lmzhw-mesa-24.0.0.drv`. Unfortunately it seems the mesa build is not fully deterministic so it's a little hard to figure out exactly what went wrong.

You can fix this by remounting `/nix/store` read-write, nuking the store path (and the main output `/nix/store/9qx6h9jhxczji2f94rdn4bjlzyw7mjb6-mesa-24.0.0/`) (might also be smart to back them up for examination later), then `nix-build --repair`ing the derivation. Make sure you have the derivation beforehand. Once that's done, reboot and make sure it all works, then run `nix-store --verify --check-contents` for good measure.

**Comment by rowanG077** [comment]
Can confirm that nuking the nix store paths and just rebuilding the issue went away. I have been using nix for ages and never seen anything like this.

**Comment by foldfree** [comment]
I had the same issue just now! weston was crashing with illegal instructions.

I removed the `mesa` folders from `/nix/store/` path (made a backup before just in case)
and then ran `sudo nix-store --repair --verify --check-contents` which succesfully re-build the mesa driver from source.
i could go back to a graphical system right after.

I noticed a lot of people are having issues related to GPU in the repo. I wonder how many could affected by this thing? Maybe it could be helpful for this to be in the maintenance/repair instructions?
